File: java/src/main/java/com/example/risk/KdbClient.java
package com.example.risk;

import com.kx.c.K;
import io.github.cdimascio.dotenv.Dotenv;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class KdbClient implements AutoCloseable {
    private static final Logger log = LoggerFactory.getLogger(KdbClient.class);
    private static final Dotenv dotenv = Dotenv.configure().ignoreIfMissing().load();
    private K connection;

    /** Resolves KDB_HOST / KDB_PORT from .env (project root) or system environment. */
    public KdbClient(String host, int port) throws Exception {
        String actualHost = host != null ? host : dotenv.get("KDB_HOST", "localhost");
        int actualPort = port > 0 ? port : Integer.parseInt(dotenv.get("KDB_PORT", "5000"));
        log.info("Connecting to kdb+ at {}:{}", actualHost, actualPort);
        connection = new K(actualHost, actualPort);
    }

    public Object query(String q) throws Exception {
        return connection.k(q);
    }

    @Override
    public void close() throws Exception {
        if (connection != null) {
            connection.close();
        }
    }
}

File: python/app/kdb_client.py
from qpython import qconnection
import os
from pathlib import Path
from dotenv import load_dotenv

# Load shared .env from project root (two levels up from this file)
load_dotenv(Path(__file__).resolve().parents[2] / ".env")

class KdbClient:
    def __init__(self, host=None, port=None):
        self.host = host or os.getenv("KDB_HOST", "localhost")
        self.port = int(port or os.getenv("KDB_PORT", "5000"))
        self.db_dir = os.getenv("KDB_DB_DIR")
        self.conn = qconnection.QConnection(host=self.host, port=self.port)
        self.conn.open()

    def query(self, q_string):
        """Send a raw q query and return the result."""
        return self.conn(q_string)

    def close(self):
        if self.conn is not None:
            self.conn.close()

    # context manager support so we can use `with KdbClient() as kdb:`
    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.close()
